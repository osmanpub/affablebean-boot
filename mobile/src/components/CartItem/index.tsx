import React, {useState} from 'react';
import {Controller, useForm} from 'react-hook-form';
import {
  ActivityIndicator,
  Button,
  Image,
  StyleSheet,
  Text,
  TextInput,
  View,
} from 'react-native';
import {connect, useDispatch} from 'react-redux';
import {getId, getProductIcon} from '../../helpers/utils';
import {Cart, CartItem as CartItemState} from '../../interfaces/cart';
import {updateProductInCart} from '../../net/cart';
import {RootState} from '../../redux';

type FormData = {
  quantity: number;
};

type Props = {
  cart: Cart;
  item: CartItemState;
};

function CartItem(props: Props) {
  const {cart, item} = props;
  const {product} = item;
  const name = product.name;

  const [visible, setVisible] = useState(true);
  const dispatch = useDispatch();
  const {control, handleSubmit, errors} = useForm<FormData>();

  const onChange = (args: Array<any>) => {
    const value = args[0].nativeEvent.text;
    setVisible(value === '' ? false : !isNaN(Number(value)));

    return {
      value,
    };
  };

  const onSubmit = handleSubmit(({quantity}) => {
    dispatch(updateProductInCart(getId(product), quantity));
  });

  return cart.isFetching ? (
    <ActivityIndicator size="large" color="blue" />
  ) : (
    <View style={styles.container}>
      <Image source={getProductIcon(name)} testID={`image-${name}`} />
      <View style={{paddingLeft: 24, flexDirection: 'column'}}>
        <Text style={styles.desc} testID={`name-${name}`}>
          {name}
        </Text>
        <Text style={styles.desc}>&euro; {item.total.toFixed(2)}</Text>
        <Text style={{paddingBottom: 8}} testID={`price-${name}`}>
          &euro; {product.price.toFixed(2)}
        </Text>
        <Controller
          as={<TextInput />}
          control={control}
          defaultValue={item.quantity.toString()}
          name="quantity"
          onChange={onChange}
          placeholder="Enter quantity"
          rules={{required: true, maxLength: 3}}
          style={styles.qty}
        />
        {errors.quantity && <Text style={styles.error}>Enter a quantity.</Text>}
        {visible && <Button onPress={onSubmit} title="update" />}
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    flexDirection: 'row',
    paddingBottom: 24,
    paddingLeft: 32,
  },
  desc: {fontWeight: 'bold'},
  error: {
    color: 'red',
    fontWeight: 'bold',
  },
  qty: {
    borderBottomColor: 'black',
    borderBottomWidth: 1,
    marginBottom: 8,
  },
});

const mapStateToProps = (state: RootState) => ({
  cart: state.cart,
});

export default connect(mapStateToProps)(CartItem);
