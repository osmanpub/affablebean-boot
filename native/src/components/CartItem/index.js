import React, { Component } from "react";
import PropTypes from "prop-types";
import { updateProductInCart } from "../../net/cart";

export default class CartItem extends Component {
  constructor(props) {
    super(props);
    const { item } = this.props;

    this.state = {
      qty: item.quantity
    };
  }

  handleChange = event => {
    const input = event.target;
    let qty = Number(input.value);

    if (qty >= 0 && qty <= 10) {
      this.setState({ qty: qty });
    } else {
      input.value = this.state.qty;
    }
  };

  handleSubmit = event => {
    event.preventDefault();
    const { dispatch, item } = this.props;
    const { product } = item;
    dispatch(updateProductInCart(product.id, Number(this.state.qty)));
  };

  render() {
    const { index, item } = this.props;
    const { product } = item;
    const name = product.name;
    const rowCol = index % 2 === 0 ? "white" : "lightBlue";
    const inputStyle = {
      margin: "5px",
      textAlign: "center"
    };

    return null;
    // return (
    //   <React.Fragment>
    //     <tr className={`${rowCol}`}>
    //       <CartTableTd>
    //         <img src={`/static/img/products/${name}.png`} alt="{name}" />
    //       </CartTableTd>
    //       <CartTableTd>{name}</CartTableTd>
    //       <CartTableTd>
    //         &euro;
    //         {item.total.toFixed(2)}
    //         <br />
    //         {product.price}
    //       </CartTableTd>
    //       <CartTableTd>
    //         <form onSubmit={this.handleSubmit}>
    //           <input
    //             className="form-control"
    //             maxLength="2"
    //             onChange={this.handleChange}
    //             size="2"
    //             style={inputStyle}
    //             type="number"
    //             value={this.state.qty}
    //           />
    //           <button className="btn btn-primary btn-sm" type="submit">
    //             update
    //           </button>
    //         </form>
    //       </CartTableTd>
    //     </tr>
    //   </React.Fragment>
    // );
  }
}

CartItem.propTypes = {
  dispatch: PropTypes.func.isRequired,
  index: PropTypes.number.isRequired,
  item: PropTypes.object.isRequired
};
