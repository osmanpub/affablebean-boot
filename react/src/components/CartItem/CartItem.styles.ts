import styled from "styled-components";

export const CartTableTd = styled.td`
  width: 145px;
  height: 90px;
`;
