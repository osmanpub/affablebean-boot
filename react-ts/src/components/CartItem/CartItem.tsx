import React, { Component } from "react";
import { updateProductInCart } from "../../net/cart";
import { CartTableTd } from "./CartItem.styles";
import { CartItemProps } from "../../interfaces/props";
import { CartItemState } from "../../interfaces/state";
import { Product } from "../../interfaces/categories";

export class CartItem extends Component<CartItemProps, CartItemState> {
  constructor(props: CartItemProps) {
    super(props);
    const { item } = this.props;

    this.state = {
      qty: item.quantity
    };
  }

  handleChange = (event: any) => {
    const input = event.target;
    let qty = Number(input.value);

    if (qty >= 0 && qty <= 10) {
      this.setState({ qty: qty });
    } else {
      input.value = this.state.qty;
    }
  };

  handleSubmit = (event: any) => {
    event.preventDefault();
    const { dispatch, item } = this.props;
    const product: Product = item.product;
    dispatch(updateProductInCart(product.id, Number(this.state.qty)));
  };

  render() {
    const { index, item } = this.props;
    const { product } = item;
    const name = product.name;
    const rowCol = index % 2 === 0 ? "white" : "lightBlue";
    const inputStyle: object = {
      margin: "5px",
      textAlign: "center"
    };

    return (
      <React.Fragment>
        <tr className={`${rowCol}`}>
          <CartTableTd>
            <img src={`/static/img/products/${name}.png`} alt="{name}" />
          </CartTableTd>
          <CartTableTd>{name}</CartTableTd>
          <CartTableTd>
            &euro;
            {item.total.toFixed(2)}
            <br />
            {product.price}
          </CartTableTd>
          <CartTableTd>
            <form onSubmit={this.handleSubmit}>
              <input
                className="form-control"
                maxLength={2}
                onChange={this.handleChange}
                size={2}
                style={inputStyle}
                type="number"
                value={this.state.qty}
              />
              <button className="`btn btn-primary btn-sm`" type="submit">
                update
              </button>
            </form>
          </CartTableTd>
        </tr>
      </React.Fragment>
    );
  }
}
