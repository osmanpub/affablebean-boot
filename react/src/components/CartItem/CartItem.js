import React, { Component } from "react";
import { updateProductInCart } from "../../net/cart";
import { CartTableTd } from "./CartItem.styles";

export class CartItem extends Component {
  constructor(props) {
    super(props);
    const { item } = this.props;

    this.state = {
      qty: item.quantity
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(event) {
    const inputQty = event.target;
    let qty = Number(inputQty.value);

    if (qty >= 0 && qty <= 10) {
      this.setState({ qty: qty });
    } else {
      inputQty.value = this.state.qty;
    }
  }

  handleSubmit(event) {
    event.preventDefault();
    const { dispatch, item } = this.props;
    const { product } = item;
    dispatch(updateProductInCart(product.id, Number(this.state.qty)));
  }

  render() {
    const { index, item } = this.props;
    const { product } = item;
    const name = product.name;
    const rowCol = index % 2 === 0 ? "white" : "lightBlue";
    const inputStyle = {
      margin: "5px",
      textAlign: "center"
    };

    return (
      <React.Fragment>
        <tr className={`${rowCol}`}>
          <CartTableTd>
            <img src={`/static/img/products/${name}.png`} alt="{name}" />
          </CartTableTd>
          <CartTableTd>{name}</CartTableTd>
          <CartTableTd>
            &euro;
            {item.total.toFixed(2)}
            <br />
            {product.price}
          </CartTableTd>
          <CartTableTd>
            <form onSubmit={this.handleSubmit}>
              <input
                className="form-control"
                maxLength="2"
                onChange={this.handleChange}
                size="2"
                style={inputStyle}
                type="number"
                value={this.state.qty}
              />
              <button className="`btn btn-primary btn-sm`" type="submit">
                update
              </button>
            </form>
          </CartTableTd>
        </tr>
      </React.Fragment>
    );
  }
}
