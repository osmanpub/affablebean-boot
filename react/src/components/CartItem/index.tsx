import React, { ChangeEvent, useState } from "react";
import { useDispatch } from "react-redux";
import { getId } from "../../helpers/utils";
import { CartItem as CartItemState } from "../../interfaces/cart";
import { updateProductInCart } from "../../net/cart";
import { CartTableTd } from "./CartItem.styles";

type Props = {
  index: number;
  item: CartItemState;
};

export default function CartItem(props: Props) {
  const { index, item } = props;
  const { product } = item;
  const [state, setState] = useState({ qty: item.quantity });
  const dispatch = useDispatch();

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    const input = event.target;
    let qty = Number(input.value);

    if (qty >= 0 && qty <= 10) {
      setState({ ...state, qty });
    } else {
      input.value = qty.toString();
    }
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    dispatch(updateProductInCart(getId(product), Number(state.qty)));
  };

  const name = product.name;
  const rowCol = index % 2 === 0 ? "white" : "lightBlue";

  return (
    <React.Fragment>
      <tr className={`${rowCol}`} data-cy={`product-${name}`}>
        <CartTableTd>
          <img src={`/static/img/products/${name}.png`} alt="{name}" />
        </CartTableTd>
        <CartTableTd>
          <span>{name}</span>
        </CartTableTd>
        <CartTableTd>
          &euro;
          {item.total.toFixed(2)}
          <br />
          {product.price}
        </CartTableTd>
        <CartTableTd>
          <form onSubmit={handleSubmit}>
            <input
              data-cy={`input-qty-${name}`}
              className="form-control"
              maxLength={2}
              onChange={handleChange}
              size={2}
              style={{ margin: "5px", textAlign: "center", width: "90%" }}
              type="number"
              value={state.qty}
            />
            <button
              className="btn btn-primary btn-sm"
              type="submit"
              data-cy={`update-qty-${name}`}
            >
              update
            </button>
          </form>
        </CartTableTd>
      </tr>
    </React.Fragment>
  );
}
