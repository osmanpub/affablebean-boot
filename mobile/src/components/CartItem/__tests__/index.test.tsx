import React from 'react';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import CartItem from '..';
import {fireEvent, render, wait} from '@testing-library/react-native';

const mockStore = configureStore([]);

const initialState = {
  cart: {
    didInvalidate: false,
    isFetching: false,
    items: [],
    numberOfItems: 0,
    subtotal: 0,
  },
};

const store = mockStore(initialState);

const item = {
  product: {
    id: 1,
    description: 'semi skimmed (1L)',
    name: 'milk',
    price: 1.7,
    category: {id: 1, name: 'dairy'},
    _links: {
      self: {href: 'http://localhost:8080/api/products/1'},
      products: {href: 'http://localhost:8080/api/products'},
    },
  },
  quantity: 1,
  total: 1.7,
};

const cartItem = (
  <Provider store={store}>
    <CartItem item={{...item}} />
  </Provider>
);

describe('<CartItem />', () => {
  it('renders correctly', () => {
    const component = renderer.create(cartItem).toJSON();
    expect(component).toMatchSnapshot();
  });

  it('show image', () => {
    const {getByTestId} = render(cartItem);
    const img = getByTestId(/image-milk/i);
    expect(img).toBeTruthy();
  });

  it('show name', () => {
    const {getByTestId} = render(cartItem);
    const name = getByTestId(/name-milk/i);
    expect(name).toBeTruthy();
  });
});
