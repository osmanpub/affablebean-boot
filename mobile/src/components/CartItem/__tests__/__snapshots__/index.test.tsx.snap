// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartItem /> renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "paddingBottom": 24,
      "paddingLeft": 32,
    }
  }
>
  <Image
    source={
      Object {
        "testUri": "../../../assets/img/products/milk.png",
      }
    }
    testID="image-milk"
  />
  <View
    style={
      Object {
        "flexDirection": "column",
        "paddingLeft": 24,
      }
    }
  >
    <Text
      style={
        Object {
          "fontWeight": "bold",
        }
      }
      testID="name-milk"
    >
      milk
    </Text>
    <Text
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      € 
      1.70
    </Text>
    <Text
      style={
        Object {
          "paddingBottom": 8,
        }
      }
      testID="price-milk"
    >
      € 
      1.70
    </Text>
    <TextInput
      allowFontScaling={true}
      name="quantity"
      onChange={[Function]}
      placeholder="Enter quantity"
      rejectResponderTermination={true}
      style={
        Object {
          "borderBottomColor": "black",
          "borderBottomWidth": 1,
          "marginBottom": 8,
        }
      }
      underlineColorAndroid="transparent"
      value="1"
    />
    <Button
      onPress={[Function]}
      title="update"
    />
  </View>
</View>
`;
